Build failure: tailscale

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

```console
$ nix build github:nixos/nixpkgs/227e660edf72cacbfdb31cf2ade2beff9bdb4a90#tailscale
```

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/291794904

### Relevant log output

```console
unpacking 'github:nixos/nixpkgs/227e660edf72cacbfdb31cf2ade2beff9bdb4a90' into the Git cache...
this derivation will be built:
  /nix/store/nm9wkgmqaz6i7fzr363ws5pmja37ijkj-tailscale-1.80.3.drv
building '/nix/store/nm9wkgmqaz6i7fzr363ws5pmja37ijkj-tailscale-1.80.3.drv'...
tailscale> Running phase: unpackPhase
tailscale> unpacking source archive /nix/store/v9zk9mm61y9n5hg0j575v7dj6bjd0w3p-source
tailscale> source root is source
tailscale> Running phase: patchPhase
tailscale> applying patch /nix/store/9yqyzxlf7mqjr29bwyldxk0xpi2hdgk1-325ca13c4549c1af58273330744d160602218af9.patch
tailscale> patching file cmd/tailscale/cli/ssh.go
tailscale> Running phase: updateAutotoolsGnuConfigScriptsPhase
tailscale> Running phase: configurePhase
tailscale> Running phase: buildPhase
tailscale> Building subPackage ./cmd/derper
tailscale> Building subPackage ./cmd/derpprobe
tailscale> Building subPackage ./cmd/tailscaled
tailscale> Building subPackage ./cmd/tsidp
tailscale> Running phase: checkPhase
tailscale> ok   tailscale.com   0.210s
tailscale> ok   tailscale.com/appc      0.167s
tailscale> ?    tailscale.com/appc/appctest     [no test files]
tailscale> ok   tailscale.com/atomicfile        0.151s
tailscale> ok   tailscale.com/chirp     0.651s
tailscale> --- FAIL: TestWhoIsPeerNotFound (0.00s)
tailscale> panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted [recovered]
tailscale>      panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted
tailscale> goroutine 5 [running]:
tailscale> testing.tRunner.func1.2({0x102b07c40, 0x14000039530})
tailscale>      /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/testing/testing.go:1632 +0x1bc
tailscale> testing.tRunner.func1()
tailscale>      /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/testing/testing.go:1635 +0x334
tailscale> panic({0x102b07c40?, 0x14000039530?})
tailscale>      /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/runtime/panic.go:785 +0x124
tailscale> net/http/httptest.newLocalListener()
tailscale>      /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/net/http/httptest/server.go:71 +0xdc
tailscale> net/http/httptest.NewUnstartedServer(...)
tailscale>      /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/net/http/httptest/server.go:119
tailscale> net/http/httptest.NewServer({0x102b951a0, 0x102b91d30})
tailscale>      /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/net/http/httptest/server.go:106 +0x28
tailscale> tailscale.com/client/tailscale.TestWhoIsPeerNotFound(0x14000115520)
tailscale>      /private/tmp/nix-build-tailscale-1.80.3.drv-0/source/client/tailscale/localclient_test.go:39 +0x3c
tailscale> testing.tRunner(0x14000115520, 0x102b91c88)
tailscale>      /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/testing/testing.go:1690 +0xe4
tailscale> created by testing.(*T).Run in goroutine 1
tailscale>      /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/testing/testing.go:1743 +0x314
tailscale> FAIL tailscale.com/client/tailscale  0.193s
tailscale> FAIL
error: builder for '/nix/store/nm9wkgmqaz6i7fzr363ws5pmja37ijkj-tailscale-1.80.3.drv' failed with exit code 1;
       last 25 log lines:
       > --- FAIL: TestWhoIsPeerNotFound (0.00s)
       > panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted [recovered]
       >         panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted
       >
       > goroutine 5 [running]:
       > testing.tRunner.func1.2({0x102b07c40, 0x14000039530})
       >        /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/testing/testing.go:1632 +0x1bc
       > testing.tRunner.func1()
       >     /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/testing/testing.go:1635 +0x334
       > panic({0x102b07c40?, 0x14000039530?})
       >       /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/runtime/panic.go:785 +0x124
       > net/http/httptest.newLocalListener()
       >   /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/net/http/httptest/server.go:71 +0xdc
       > net/http/httptest.NewUnstartedServer(...)
       >     /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/net/http/httptest/server.go:119
       > net/http/httptest.NewServer({0x102b951a0, 0x102b91d30})
       >    /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/net/http/httptest/server.go:106 +0x28
       > tailscale.com/client/tailscale.TestWhoIsPeerNotFound(0x14000115520)
       >  /private/tmp/nix-build-tailscale-1.80.3.drv-0/source/client/tailscale/localclient_test.go:39 +0x3c
       > testing.tRunner(0x14000115520, 0x102b91c88)
       >        /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/testing/testing.go:1690 +0xe4
       > created by testing.(*T).Run in goroutine 1
       >   /nix/store/yjqrp3d93mxa2g80j57whyk55xwxgl7r-go-1.23.6/share/go/src/testing/testing.go:1743 +0x314
       > FAIL tailscale.com/client/tailscale  0.193s
       > FAIL
       For full logs, run 'nix log /nix/store/nm9wkgmqaz6i7fzr363ws5pmja37ijkj-tailscale-1.80.3.drv'.
```

### Additional context

Possibly related to re-enabling tests? https://github.com/NixOS/nixpkgs/pull/386492/commits/079c28fb814d4f6355e8a1e4a59c340d6eb3ca3e

Perhaps needs a __darwinAllowLocalNetwork or whatever that setting is. Will try to test soon.

### System metadata

 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.3.0, macOS 15.3.1`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.12`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/store/vjbklmp9rjkq8cw7kl4qbi6l71kjdyf3-source`



### Notify maintainers

@mbaillie
@jk
@mfrw
@pyrox0
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

Still fails with `__darwinAllowLocalNetworking = true;`, but the failure is different:

```
this derivation will be built:
  /nix/store/f4bk0fd68ghvdnhd2q8m9mqcb51k0jyr-tailscale-1.80.3.drv
building '/nix/store/f4bk0fd68ghvdnhd2q8m9mqcb51k0jyr-tailscale-1.80.3.drv'...
tailscale> Running phase: unpackPhase
tailscale> unpacking source archive /nix/store/v9zk9mm61y9n5hg0j575v7dj6bjd0w3p-source
tailscale> source root is source
tailscale> Running phase: patchPhase
tailscale> applying patch /nix/store/9yqyzxlf7mqjr29bwyldxk0xpi2hdgk1-325ca13c4549c1af58273330744d160602218af9.patch
tailscale> patching file cmd/tailscale/cli/ssh.go
tailscale> Running phase: updateAutotoolsGnuConfigScriptsPhase
tailscale> Running phase: configurePhase
tailscale> Running phase: buildPhase
tailscale> Building subPackage ./cmd/derper
tailscale> Building subPackage ./cmd/derpprobe
tailscale> Building subPackage ./cmd/tailscaled
tailscale> Building subPackage ./cmd/tsidp
tailscale> Running phase: checkPhase
tailscale> ok   tailscale.com   0.193s
tailscale> ok   tailscale.com/appc      0.160s
tailscale> ?    tailscale.com/appc/appctest     [no test files]
tailscale> ok   tailscale.com/atomicfile        0.149s
tailscale> ok   tailscale.com/chirp     0.641s
tailscale> ok   tailscale.com/client/tailscale  0.301s
tailscale> ok   tailscale.com/client/web        0.228s
tailscale> ok   tailscale.com/clientupdate      0.183s
tailscale> ok   tailscale.com/clientupdate/distsign     0.197s
tailscale> ok   tailscale.com/cmd/cloner        0.146s
tailscale> package tailscale.com/cmd/containerboot: build constraints exclude all Go files in /private/tmp/nix-build-tailscale-1.80.3.drv-0/source/cmd/containerboot
tailscale> ok   tailscale.com/cmd/derper        0.512s
tailscale> ok   tailscale.com/cmd/gitops-pusher 0.154s
tailscale> ok   tailscale.com/cmd/k8s-nameserver        0.316s
tailscale> ok   tailscale.com/cmd/k8s-operator  0.557s
tailscale> ok   tailscale.com/cmd/k8s-operator/e2e      0.460s
tailscale> ok   tailscale.com/cmd/k8s-operator/generate 0.142s [no tests to run]
tailscale> ok   tailscale.com/cmd/sniproxy      1.887s
tailscale> ?    tailscale.com/cmd/speedtest     [no test files]
tailscale> ok   tailscale.com/cmd/tailscale     0.350s
tailscale> ok   tailscale.com/cmd/tailscale/cli 0.626s
tailscale> ok   tailscale.com/cmd/tailscale/cli/ffcomplete/internal     0.153s
tailscale> ok   tailscale.com/cmd/tailscaled    1.181s
tailscale> ok   tailscale.com/cmd/testwrapper   1.776s
tailscale> ok   tailscale.com/cmd/testwrapper/flakytest 0.122s
tailscale> ok   tailscale.com/cmd/viewer        0.163s
tailscale> ok   tailscale.com/control/controlbase       0.274s
tailscale> ok   tailscale.com/control/controlclient     0.224s
tailscale> ok   tailscale.com/control/controlhttp       0.317s
tailscale> ok   tailscale.com/control/controlknobs      0.142s
tailscale> ok   tailscale.com/derp      0.571s
tailscale> ok   tailscale.com/derp/derphttp     0.937s
tailscale> ?    tailscale.com/derp/xdp  [no test files]
tailscale> ok   tailscale.com/disco     0.149s
tailscale> ok   tailscale.com/doctor    0.172s
tailscale> ok   tailscale.com/doctor/permissions        0.142s
tailscale> ok   tailscale.com/drive     0.154s
tailscale> ok   tailscale.com/drive/driveimpl   7.293s
tailscale> ok   tailscale.com/drive/driveimpl/compositedav      0.713s
tailscale> ok   tailscale.com/drive/driveimpl/dirfs     0.205s
tailscale> ok   tailscale.com/drive/driveimpl/shared    0.158s
tailscale> ?    tailscale.com/envknob   [no test files]
tailscale> ok   tailscale.com/envknob/logknob   0.147s
tailscale> ok   tailscale.com/health    1.192s
tailscale> ok   tailscale.com/hostinfo  0.159s
tailscale> ok   tailscale.com/ipn       0.273s
tailscale> ok   tailscale.com/ipn/ipnlocal      5.455s
tailscale> ok   tailscale.com/ipn/ipnserver     0.336s
tailscale> ok   tailscale.com/ipn/localapi      0.258s
tailscale> ok   tailscale.com/ipn/store 0.186s
tailscale> ?    tailscale.com/ipn/store/awsstore        [no test files]
tailscale> ok   tailscale.com/jsondb    0.165s
tailscale> ok   tailscale.com/k8s-operator      0.222s
tailscale> ok   tailscale.com/k8s-operator/sessionrecording     0.287s
tailscale> ok   tailscale.com/k8s-operator/sessionrecording/spdy        0.258s
tailscale> ok   tailscale.com/k8s-operator/sessionrecording/ws  0.226s
tailscale> ok   tailscale.com/kube/egressservices       0.154s
tailscale> ok   tailscale.com/kube/kubeclient   0.174s
tailscale> ok   tailscale.com/log/filelogger    0.135s
tailscale> ok   tailscale.com/log/sockstatlog   0.204s
tailscale> ok   tailscale.com/logpolicy 0.164s
tailscale> ok   tailscale.com/logtail   2.183s
tailscale> ok   tailscale.com/logtail/filch     0.151s
tailscale> ok   tailscale.com/metrics   0.153s
tailscale> ok   tailscale.com/net/art   0.496s
tailscale> ok   tailscale.com/net/bakedroots    0.149s
tailscale> ok   tailscale.com/net/captivedetection      0.181s
tailscale> ok   tailscale.com/net/connstats     0.191s
tailscale> ok   tailscale.com/net/dns   0.218s
tailscale> ok   tailscale.com/net/dns/publicdns 0.136s
tailscale> ok   tailscale.com/net/dns/recursive 0.161s
tailscale> ok   tailscale.com/net/dns/resolvconffile    0.127s
tailscale> ok   tailscale.com/net/dns/resolver  0.236s
tailscale> ok   tailscale.com/net/dnscache      0.155s
tailscale> ok   tailscale.com/net/dnsfallback   0.192s
tailscale> ok   tailscale.com/net/flowtrack     0.143s
tailscale> ok   tailscale.com/net/ipset 0.148s
tailscale> ok   tailscale.com/net/ktimeout      0.127s
tailscale> ok   tailscale.com/net/memnet        0.460s
tailscale> ok   tailscale.com/net/netcheck      6.289s
tailscale> ?    tailscale.com/net/neterror      [no test files]
tailscale> --- FAIL: TestLikelyHomeRouterIPSyscallExec (0.00s)
tailscale>     interfaces_darwin_test.go:22: syscall() = 192.168.8.1, true, netstat = invalid IP, false
tailscale>     interfaces_darwin_test.go:38: syscall default route interface en0 differs from netstat
tailscale> FAIL
tailscale> FAIL tailscale.com/net/netmon        0.169s
tailscale> FAIL
error: builder for '/nix/store/f4bk0fd68ghvdnhd2q8m9mqcb51k0jyr-tailscale-1.80.3.drv' failed with exit code 1;
       last 25 log lines:
       > ok    tailscale.com/logtail/filch     0.151s
       > ok      tailscale.com/metrics   0.153s
       > ok      tailscale.com/net/art   0.496s
       > ok      tailscale.com/net/bakedroots    0.149s
       > ok      tailscale.com/net/captivedetection      0.181s
       > ok      tailscale.com/net/connstats     0.191s
       > ok      tailscale.com/net/dns   0.218s
       > ok      tailscale.com/net/dns/publicdns 0.136s
       > ok      tailscale.com/net/dns/recursive 0.161s
       > ok      tailscale.com/net/dns/resolvconffile    0.127s
       > ok      tailscale.com/net/dns/resolver  0.236s
       > ok      tailscale.com/net/dnscache      0.155s
       > ok      tailscale.com/net/dnsfallback   0.192s
       > ok      tailscale.com/net/flowtrack     0.143s
       > ok      tailscale.com/net/ipset 0.148s
       > ok      tailscale.com/net/ktimeout      0.127s
       > ok      tailscale.com/net/memnet        0.460s
       > ok      tailscale.com/net/netcheck      6.289s
       > ?       tailscale.com/net/neterror      [no test files]
       > --- FAIL: TestLikelyHomeRouterIPSyscallExec (0.00s)
       >     interfaces_darwin_test.go:22: syscall() = 192.168.8.1, true, netstat = invalid IP, false
       >     interfaces_darwin_test.go:38: syscall default route interface en0 differs from netstat
       > FAIL
       > FAIL      tailscale.com/net/netmon        0.169s
       > FAIL
       For full logs, run 'nix log /nix/store/f4bk0fd68ghvdnhd2q8m9mqcb51k0jyr-tailscale-1.80.3.drv'.
```

--- a/pkgs/by-name/ta/tailscale/package.nix
+++ b/pkgs/by-name/ta/tailscale/package.nix
@@ -16,6 +16,8 @@
   procps,
   # runtime tooling - darwin
   lsof,
+  # check phase tooling - darwin
+  unixtools,
 
   nixosTests,
   tailscale-nginx-auth,
@@ -56,6 +58,10 @@ buildGoModule {
     installShellFiles
   ];
 
+  nativeCheckInputs = lib.optionals stdenv.hostPlatform.isDarwin [
+    unixtools.netstat
+  ];
+
   env.CGO_ENABLED = 0;
 
   subPackages = [
@@ -88,6 +94,10 @@ buildGoModule {
     rm -rf ./tool
   '';
 
+  # Tests start http servers which need to bind to local addresses:
+  # panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted
+  __darwinAllowLocalNetworking = true;
+
   preCheck = ''
     # feed in all tests for testing
     # subPackages above limits what is built to just what we
@@ -100,39 +110,51 @@ buildGoModule {
 
   checkFlags =
     let
-      skippedTests = [
-        # dislikes vendoring
-        "TestPackageDocs" # .
-        # tries to start tailscaled
-        "TestContainerBoot" # cmd/containerboot
+      skippedTests =
+        [
+          # dislikes vendoring
+          "TestPackageDocs" # .
+          # tries to start tailscaled
+          "TestContainerBoot" # cmd/containerboot
 
-        # just part of a tool which generates yaml for k8s CRDs
-        # requires helm
-        "Test_generate" # cmd/k8s-operator/generate
-        # self reported potentially flakey test
-        "TestConnMemoryOverhead" # control/controlbase
+          # just part of a tool which generates yaml for k8s CRDs
+          # requires helm
+          "Test_generate" # cmd/k8s-operator/generate
+          # self reported potentially flakey test
+          "TestConnMemoryOverhead" # control/controlbase
 
-        # interacts with `/proc/net/route` and need a default route
-        "TestDefaultRouteInterface" # net/netmon
-        "TestRouteLinuxNetlink" # net/netmon
-        "TestGetRouteTable" # net/routetable
+          # interacts with `/proc/net/route` and need a default route
+          "TestDefaultRouteInterface" # net/netmon
+          "TestRouteLinuxNetlink" # net/netmon
+          "TestGetRouteTable" # net/routetable
 
-        # remote udp call to 8.8.8.8
-        "TestDefaultInterfacePortable" # net/netutil
+          # remote udp call to 8.8.8.8
+          "TestDefaultInterfacePortable" # net/netutil
 
-        # launches an ssh server which works when provided openssh
-        # also requires executing commands but nixbld user has /noshell
-        "TestSSH" # ssh/tailssh
-        # wants users alice & ubuntu
-        "TestMultipleRecorders" # ssh/tailssh
-        "TestSSHAuthFlow" # ssh/tailssh
-        "TestSSHRecordingCancelsSessionsOnUploadFailure" # ssh/tailssh
-        "TestSSHRecordingNonInteractive" # ssh/tailssh
+          # launches an ssh server which works when provided openssh
+          # also requires executing commands but nixbld user has /noshell
+          "TestSSH" # ssh/tailssh
+          # wants users alice & ubuntu
+          "TestMultipleRecorders" # ssh/tailssh
+          "TestSSHAuthFlow" # ssh/tailssh
+          "TestSSHRecordingCancelsSessionsOnUploadFailure" # ssh/tailssh
+          "TestSSHRecordingNonInteractive" # ssh/tailssh
 
-        # test for a dev util which helps to fork golang.org/x/crypto/acme
-        # not necessary and fails to match
-        "TestSyncedToUpstream" # tempfork/acme
-      ];
+          # test for a dev util which helps to fork golang.org/x/crypto/acme
+          # not necessary and fails to match
+          "TestSyncedToUpstream" # tempfork/acme
+        ]
+        ++ lib.optionals stdenv.hostPlatform.isDarwin [
+          # syscall default route interface en0 differs from netstat
+          "TestLikelyHomeRouterIPSyscallExec" # net/netmon
+
+          # Even with __darwinAllowLocalNetworking this doesn't work.
+          # panic: write udp [::]:59507->127.0.0.1:50830: sendto: operation not permitted
+          "TestUDP" # net/socks5
+
+          # portlist_test.go:81: didn't find ephemeral port in p2 53643
+          "TestPoller" # portlist
+        ];
     in
     [ "-skip=^${builtins.concatStringsSep "$|^" skippedTests}$" ];
 

